### PR TITLE
Properly Watch() the processor image ConfigMap

### LINKED ISSUE #209: Processor controller Tracks() CM but doesn't Watch() it

### PR DESCRIPTION
**Pull request by ericbottard**
Fixes #209

Without this fix, changing the value in the ConfigMap did not trigger
a change in the deployed processors.

### CODE REVIEW

File: pkg/controllers/streaming/processor_controller.go [hunk 1]
```diff
@@ -682,6 +684,21 @@ func (r *ProcessorReconciler) SetupWithManager(mgr ctrl.Manager) error {
 			}),
 		}
 	}
+	enqueueNamedTrackedConfigMap := &handler.EnqueueRequestsFromMapFunc{
```
Comment (scothis) [inline pkg/controllers/streaming/processor_controller.go]: Let's combine this with enqueueTrackedResource similar to what @joshrider came up with:
https://github.com/projectriff/system/blob/28aac56da45362d4852822be5f93c2825a539533/pkg/controllers/core/deployer_controller.go#L614-L635

Comment (joshrider) [reply pkg/controllers/streaming/processor_controller.go]: credit: @andrew-su

Comment (ericbottard) [reply pkg/controllers/streaming/processor_controller.go]: Not sure I get what you mean. Are you asking for me to *also* replace `enqueueTrackedResources` that already exists above with the one referenced (and keep `enqueueNamedTrackedConfigMap`) or to come up with a unique function that covers both cases (maybe with an optional name for example)?

Comment (scothis) [reply pkg/controllers/streaming/processor_controller.go]: we can replace both with the linked method.

> maybe with an optional name for example

I don't follow

Comment (ericbottard) [reply pkg/controllers/streaming/processor_controller.go]: The one I added has `&& a.Meta.GetName() == processorImages` in the condition.

Comment (scothis) [reply pkg/controllers/streaming/processor_controller.go]: ah yes. That's a performance optimization, but is still correct without that check.

I'd like to eventually pull this method into a separate utility so we don't need to repeat it everywhere. If you don't make that change in this PR, we can pick it up then.

Comment (ericbottard) [reply pkg/controllers/streaming/processor_controller.go]: OK, changed to use that one function.

Notice how I've moved the 
```go
		versionKinds, _, err := r.Scheme.ObjectKinds(t)
		if err != nil {
			panic(err)
		}
```
bit out of the callback though.

File: pkg/controllers/streaming/processor_controller.go [hunk 2]
```diff
@@ -667,12 +669,16 @@ func (*ProcessorReconciler) collectAliases(bindings []streamingv1alpha1.StreamBi
...
+		if err != nil {
+			panic(err)
+		}
 		return &handler.EnqueueRequestsFromMapFunc{
 			ToRequests: handler.ToRequestsFunc(func(a handler.MapObject) []reconcile.Request {
-				requests := []reconcile.Request{}
+				var requests []reconcile.Request
 				key := tracker.NewKey(
-					t.GetGroupVersionKind(),
+					versionKinds[0],
```
Comment (ericbottard) [inline pkg/controllers/streaming/processor_controller.go]: Maybe we should also panic above if len(versionKinds) != 1

Comment (scothis) [reply pkg/controllers/streaming/processor_controller.go]: > Maybe we should also panic above if len(versionKinds) != 1

There can be multiple valid GroupVersions for a single type. Ingress is in `networking.k8s.io/v1beta1 ` and `extensions/v1beta1`.

**Comment by ericbottard** [comment]
Merged as 1377163

**Pull request closed without merging**